### PR TITLE
Improve 2fa: ask for code before account removal and 2fa disabling

### LINKED ISSUE #3816: Ask for password before account deletion
### Issue Summary
Currently a logged in user can delete it's account without asking for credentials again before confirming so.

### PR DESCRIPTION
**Pull request by agustif**
## What does this PR do?

This PR is focused around hardening the security of accounts at cal.com, we aim to do so by disallowing account removal without providing also a two factor code for confirmation if user has 2FA enabled.

We also ask for the code to disable 2FA.

### `settings/profile`
- [x] ask for password and or totpCode before account removal
### `packages/trpc/server/routers/viewer` 
- [x] add more checks to deleteMe mutation
- [x] Asks for both password and totpCode as input
- [x] Finds user by email, and adds all same checks than login (user exists, valid pass, 2fa code) before deletion.

### TwoFactor.tsx
- [x] Adds Label `"2fa_code"`
- [x] Adds new center prop (Defaults to true)

### TwoFactorModalHeader.tsx
- [x] Fix icon invisible color to white (was black on black before)

### EnableTwoFactorModal.tsx
- [x] use `<TwoFactor />` instead of `<input>` for better UX
- [x] use `Form` from react-hooks-form as it's required for our TwoFactor component

### DisableTwoFactorModal.tsx
- [x] use `<TwoFactor />` instead of `<input>` for better UX
- [x] Wrapped on `<Form>` as it's required for our own TwoFactor component
- [x] use `<PasswordField />` instead of `<input type="password">`

### `apps/web/pages/api/me` DELETE user endpoint
- [x] deleted / deprecated in favor of new trpc mutations

### TwoFactorAuth.API
- make disable() required password and 2FA-code
- make `api/auth/two-factor/disable` make all same checks than login for valid 2FA disable.

### @lib/auth -> @calcom/lib/auth
- [x] marked @lib/auth functions as deprecated 

## Translations
- [ ] `"2fa_required": "Two factor authentication required"`
- [ ] `"incorrect_2fa": "Incorrect two factor authentication code"`

**Environment**: Staging(main branch) / Production

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## How should this be tested?
Use any account to test the following:
- [ ] Enable 2fa in your account
- [ ] Try to delete account, should ask both for password and 2FA-code
- [ ] Try to disable 2FA, should ask for both too.

### Not tackled in this PR but related

- [ ] Update all `@lib/auth -> @calcom/lib/auth`, then remove `apps/web/lib/auth.ts`
- [ ] Deprecate TwoFactorAuth.API in favor of a trpc.viewer/mutations for TwoFactor.

### CODE REVIEW

**Comment by agustif** [comment]
Fixed delete account playwright test *hopefully* it passes now, basically filling password before submit

File: apps/web/components/security/EnableTwoFactorModal.tsx [hunk 1]
```diff
@@ -159,24 +160,20 @@ const EnableTwoFactorModal = ({ onEnable, onCancel }: EnableTwoFactorModalProps)
...
               <label htmlFor="code" className="mt-4 block text-sm font-medium text-gray-700">
                 {t("code")}
               </label>
               <div className="mt-1">
-                <input
-                  type="text"
-                  name="code"
-                  id="code"
-                  required
+                <OtpInput
```
Comment (zomars) [inline apps/web/components/security/EnableTwoFactorModal.tsx:169]: Why not use the [one we already have](https://github.com/calcom/cal.com/blob/main/apps/web/components/auth/TwoFactor.tsx#L9)?

File: apps/web/components/security/EnableTwoFactorModal.tsx [hunk 2]
```diff
@@ -159,24 +160,20 @@ const EnableTwoFactorModal = ({ onEnable, onCancel }: EnableTwoFactorModalProps)
...
+                  autocomplete="code"
```
Comment (zomars) [inline apps/web/components/security/EnableTwoFactorModal.tsx:170]:
```suggestion
                  autoComplete="one-time-code"
```

File: apps/web/package.json [hunk 3]
```diff
@@ -105,6 +105,7 @@
     "react-intl": "^5.25.1",
     "react-live-chat-loader": "^2.7.3",
     "react-multi-email": "^0.5.3",
+    "react-otp-input": "^2.4.0",
```
Comment (zomars) [inline apps/web/package.json:108]: Same comment as above

**Review by zomars — CHANGES_REQUESTED** [review]

File: apps/web/components/security/EnableTwoFactorModal.tsx (hunk 1 above)
Comment (agustif) [reply apps/web/components/security/EnableTwoFactorModal.tsx:169]: I tried to use it at first, but had some hard deps on using rhf with it, and I was not using a form already?

Should I move it to the one we already have and add the necessary forms if needed?

Comment (zomars) [reply apps/web/components/security/EnableTwoFactorModal.tsx:169]: Yes, let's keep it consistent regardless of what we're using

Comment (agustif) [reply apps/web/components/security/EnableTwoFactorModal.tsx]: Done, now we reuse TwoFactor and added react hook forms on were we use it

File: apps/web/components/security/EnableTwoFactorModal.tsx [hunk 4]
```diff
@@ -1,10 +1,14 @@
-import React, { SyntheticEvent, useState } from "react";
+import React, { BaseSyntheticEvent, FormEvent, useState } from "react";
```
Comment (zomars) [inline apps/web/components/security/EnableTwoFactorModal.tsx:1]:
```suggestion
import React, { BaseSyntheticEvent, useState } from "react";
```

File: apps/web/components/security/EnableTwoFactorModal.tsx [hunk 5]
```diff
@@ -1,10 +1,14 @@
...
+import { useForm } from "react-hook-form";
 
+import { useLocale } from "@calcom/lib/hooks/useLocale";
 import Button from "@calcom/ui/Button";
 import { Dialog, DialogContent } from "@calcom/ui/Dialog";
+import { Label, Form } from "@calcom/ui/v2/core/form/fields";
```
Comment (zomars) [inline apps/web/components/security/EnableTwoFactorModal.tsx:7]:
```suggestion
import { Form } from "@calcom/ui/v2/core/form/fields";
```

File: packages/trpc/server/routers/viewer.tsx [hunk 6]
```diff
@@ -30,6 +33,8 @@ import slugify from "@calcom/lib/slugify";
 import prisma, { baseEventTypeSelect, bookingMinimalSelect } from "@calcom/prisma";
 import { resizeBase64Image } from "@calcom/web/server/lib/resizeBase64Image";
 
+import { ErrorCode, verifyPassword } from "@lib/auth";
```
Comment (zomars) [inline packages/trpc/server/routers/viewer.tsx:36]: We cannot use `@lib` outside `apps/web`

File: apps/web/pages/api/user/me.ts [hunk 7]
```diff
@@ -1,42 +0,0 @@
-import type { NextApiRequest, NextApiResponse } from "next";
-
-import { deleteStripeCustomer } from "@calcom/app-store/stripepayment/lib/customer";
-import prisma from "@calcom/prisma";
-
-import { getSession } from "@lib/auth";
-
-export default async function handler(req: NextApiRequest, res: NextApiResponse) {
```
Comment (zomars) [inline apps/web/pages/api/user/me.ts:8]: Are we sure this is not being used anymore?

Comment (agustif) [reply apps/web/pages/api/user/me.ts:8]: I couldn't find anywhere else where it was being called besides the delete account feature, so I'm pretty sure yeah

File: packages/trpc/server/routers/viewer.tsx (hunk 6 above)
Comment (agustif) [reply packages/trpc/server/routers/viewer.tsx:36]: Should I move the ones I need to re-use in trpc to new packages/auth or just packages/lib/auth.ts?

**Comment by agustif** [comment]
Not sure why test-e2e are breaking @zomars

**Comment by zomars** [comment]
Adding the "tests needed" tag. @alannnc can you provide some quick E2E tests for 2FA enabling and disabling?

**Review by zomars — APPROVED** [review]
Thank you @agustif 🙏🏽

**Comment by agustif** [comment]
> Thank you Agusti Fernandez Pardo 🙏🏽

Thank YOU @zomars! for all the help and feedback to ship the best code possible as always ;)

Appreciate all your guidance.

Fixed conflicts!

**Comment by PeerRich** [comment]
still failing :(

File: apps/web/playwright/auth/delete-account.test.ts [hunk 8]
```diff
@@ -11,6 +11,7 @@ test("Can delete user account", async ({ page, users }) => {
   await page.waitForSelector("[data-testid=dashboard-shell]");
 
   await page.goto(`/settings/profile`);
+  await page.fill("[data-testid=password]", "delete-me");
```
Comment (zomars) [inline apps/web/playwright/auth/delete-account.test.ts:14]: - This is trying to fill the password before the modal opens
- The password is incorrect, it is generated per-test

**Pull request merged**